enhancements/ota/distribute-secondary-metadata-as-container-image.md

```diff
@@ -0,0 +1,165 @@
+
+### Repository location and name
+Under which well-known container image repository will we publish the images?
+
+
+### Continuous build and release
+Can CPaaS handle continuous build and release from master?
+
+## Summary
+The [cincinnati-graph-data GitHub repository][cincinnati-graph-data] will be continuously packaged and released in form of a container image by the means of integration with CPaaS.
+This image will be available for public consumption, and will be used by the publicly available Cincinnati instances operated by Red Hat.
+The image can be transferred and served to Cincinnati instances operated by any party, including but not limited to network-restricted environments.
+The source of the container image, namely the cincinnati-graph-data, will remain publicly available and its workflows unaffected by this enhancement.
+
+## Motivation
+
+### Goals
+- The container image will become the primary way of consuming and distributing the content of the cincinnati-graph-data repository.
+- As long as the primary release metadata will be packaged, distributed and consumed in form of container images as well, this streamlines all transport mechanisms required for data consumption by Cincinnati, as well as mirroring and serving its consumed content.
+- Every change to the master branch of the cincinnati-graph-data results in a new container image containing the unmodified working tree files of the new revision.
+- The image will be pushed to a publicly available container image registry und a well-known name.
+
+### Non-Goals
+- Changes to the GitHub repository workflow or data format and make progress.
+- Implementation details about the streamlining of the mirroring process for primary and secondary release metadata.
+
+## Proposal
+
+### User Stories [optional]
+
+#### Content mirroring for Cincinnati deployments
+A user might want to mirror all content which is required by Cincinnati.
+
+1. The user sets up a container image registry in the target environment.
+2. Optionally, if the target environment does not have internet access, the user downloads the release container images and the desired graph-data container image to a transfer medium.
+3. The user pushes the release and graph-data images to the target container image registry
+4. The user can now configure the Cincinnati instance in the target environment to consume the content from the available container image registry
+
+### Implementation Details/Notes/Constraints [optional]
+* CPaaS needs to be configured
+
+### Risks and Mitigations
+
+### Container image content integrity
+The content of the cincinnati-graph-data repository may be altered by the process of being packaged by CPaaS and distributed via the container image registry.
+
+#### Mitigation
+A mitigation strategy is to sign the resulting image in the CPaaS pipeline, publish the signatures, and have Cincinnati verify the image after download.
```

Comment:
Maybe just "have Cincinnati verify the signature.", because ideally the sig-check happens before downloading much, to avoid tarbombs and huge-layer DoS sorts of things.

What is the plan for teaching Cincinnati about trusted keys?  Do we have a config setting where we inline an ASCII-armored GPG key that should be trusted?  And then reboot Cincinnati when rotating keys?  Teach Cincinnati to be able to configure multiple trusted signers to allow more graceful key rotation?  Which image-signature-distribution mechanisms do we support?

Reply:
> Maybe just "have Cincinnati verify the signature.", because ideally the sig-check happens before downloading much, to avoid tarbombs and huge-layer DoS sorts of things.

We want to ensure that the image is signed, which is what the previous part of the sentence is about. In my understanding, signature checking involves verifying the content integrity. Are you suggesting to download and verify the signatures first by themselves, and only on success proceed to download and verify the content?

> What is the plan for teaching Cincinnati about trusted keys?

There's no plan, so we can discuss it here. Do you think adding the results under the mitigation section is appropriate?

> Do we have a config setting where we inline an ASCII-armored GPG key that should be trusted? And then reboot Cincinnati when rotating keys? Teach Cincinnati to be able to configure multiple trusted signers to allow more graceful key rotation?

Having an an array of ASCII-armored GPG pubkeys sounds like a viable plan. This would be a plugin setting.

> Which image-signature-distribution mechanisms do we support?

I was thinking about shipping them in a container image as well. Just to be sure, it would be a single signature, right?

Reply:
> Are you suggesting to download and verify the signatures first by themselves, and only on success proceed to download and verify the content?

Yeah.  The signature [asserts a pullspec-digest association][1] (e.g. "sha256:817a12c32a39bbe394944ba49de563e085f1d3c5266eb8e9723256bc4448680e is a valid digest for docker.io/library/busybox:latest").  If you lookup by tag, the tag also provides a pullspec-digest association.  If you don't find any signatures you trust asserting the pullspec-digest pair claimed by the tag, then there's no need to download anything further.  If you find a signature you trust that asserts the pullspec-digest pair claimed by the tag, you can safely proceed to download the by-digest image.  And your pulling library needs to ensure that downloaded content matches the claimed digests (e.g. that the top-level manifest for the image hashes to sha256:817a12...), to safely transfer the trust established by the valid signature.

> Do you think adding the results under the mitigation section is appropriate?

I think there will need to be something in the proposal section about how users will tell Cincinnati which signers they trust, and about which protocol(s) Cincinnati will use to look up signatures.  Then the mitigation section can reference that portion of the proposal and say:

```markdown
We're mitigating this risk [with image signatures](#link-to-proposal-section-about-signatures).
```

> Having an an array of ASCII-armored GPG pubkeys sounds like a viable plan.

Might be easier to skip the array and accept an ASCII-armored keyring, since a keyring can already contain multiple keys.

> I was thinking about shipping them in a container image as well.

Not possible, because the signature includes the hash of the image's top-level manifest, and unless you've broken the hash you cannot have a Merkle loop like:

1. Image layer, including a signature signing the manifest hash from (2).
2. Manifest, including a hash of the layer from (1).

> Just to be sure, it would be a single signature, right?

Other schemes are possible (e.g. require both ops and prod-sec signatures).  Formats like [containers-policy.json(5)][2] allow folks to declare policies like that.  For Red Hat and Cincinnati though, I think "we can find at least one signature that is trusted by at least one of the signers in the configured keyring" is probably enough, at least for an initial implementation.

[1]: https://github.com/containers/image/blob/04c2ec3e6dedaf1b680753c671258180e59d9bbf/docs/containers-signature.5.md#json-data-format
[2]: https://github.com/containers/image/blob/04c2ec3e6dedaf1b680753c671258180e59d9bbf/docs/containers-policy.json.5.md

Reply:
> Not possible, because the signature includes the hash of the image's top-level manifest, and unless you've broken the hash you cannot have a Merkle loop like:

I think it's possible. What I mean is have the signatures in a separate repository, while you're discussing having them in a subsequent layer of an image in the same repository.

Reply:
> Yeah. The signature [asserts a pullspec-digest association](https://github.com/containers/image/blob/04c2ec3e6dedaf1b680753c671258180e59d9bbf/docs/containers-signature.5.md#json-data-format) ...

That doesn't answer my question. I tried to ask you if you think that the trust towards the signing certificate should be verified before even proceeding to attempt to verify content integrity.



> Might be easier to skip the array and accept an ASCII-armored keyring, since a keyring can already contain multiple keys.

I'm not expert here, but a keyring sounds like a more complex data structure than an array of keys. Do you think it's important to lay this implementation detail out here? I'm fine if it his but I'd be surprised.

Reply:
Use of keyring may make it easier to implement verify by re-using logic (porting to Rust I suppose) already in use in oc and cvo (https://github.com/openshift/cluster-version-operator/blob/e318b86d674ccc850266b74169a0a4403d3b633b/pkg/verify/verify.go#L227-L241). Not sure but something to consider.

Reply:
I mentioned the keyring as a configuration option to the new plugin. I marked it as optional to remain flexible to environments where the image is not signed.